### PR TITLE
revert #6595 #6820

### LINKED ISSUE #6819: Optimizer rule 'push_down_projection' failed
### Describe the bug

This problem was found in ballista's test: https://github.com/apache/arrow-ballista/actions/runs/5430821519/jobs/9876790266?pr=834

```sql
❯ select l_returnflag, sum(l_extendedprice * 1) as sum_disc_price
            from lineitem
            group by l_returnflag
            order by l_returnflag;
push_down_projection
caused by
Internal error: Optimizer rule 'push_down_projection' failed, due to generate a different schema, original schema: DFSchema { fields: [DFField { qualifier: Some(Bare { table: "lineitem" }), field: Field { name: "l_returnflag", data_type: Utf8, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} } }, DFField { qualifier: None, field: Field { name: "sum_disc_price", data_type: Decimal128(38, 2), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }], metadata: {} }, new schema: DFSchema { fields: [DFField { qualifier: Some(Bare { table: "lineitem" }), field: Field { name: "l_returnflag", data_type: Utf8, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} } }, DFField { qualifier: None, field: Field { name: "sum_disc_price", data_type: Decimal128(25, 2), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }], metadata: {} }. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
```

### To Reproduce

datafusion = "27.0.0"

```shell
❯ CREATE EXTERNAL TABLE lineitem
STORED AS PARQUET
LOCATION '/Users/r4ntix/Workspace/github/arrow-datafusion/benchmarks/tpch-parquet/lineitem.parquet/part-0.parquet';
0 rows in set. Query took 0.006 seconds.

❯ describe lineitem;
+-----------------+-------------------+-------------+
| column_name     | data_type         | is_nullable |
+-----------------+-------------------+-------------+
| l_orderkey      | Int64             | NO          |
| l_partkey       | Int64             | NO          |
| l_suppkey       | Int64             | NO          |
| l_linenumber    | Int32             | NO          |
| l_quantity      | Decimal128(15, 2) | NO          |
| l_extendedprice | Decimal128(15, 2) | NO          |
| l_discount      | Decimal128(15, 2) | NO          |
| l_tax           | Decimal128(15, 2) | NO          |
| l_returnflag    | Utf8              | NO          |
| l_linestatus    | Utf8              | NO          |
| l_shipdate      | Date32            | NO          |
| l_commitdate    | Date32            | NO          |
| l_receiptdate   | Date32            | NO          |
| l_shipinstruct  | Utf8              | NO          |
| l_shipmode      | Utf8              | NO          |
| l_comment       | Utf8              | NO          |
+-----------------+-------------------+-------------+
16 rows in set. Query took 0.001 seconds.

❯ select l_returnflag, sum(l_extendedprice * 1) as sum_disc_price
            from lineitem
            group by l_returnflag
            order by l_returnflag;
push_down_projection
caused by
Internal error: Optimizer rule 'push_down_projection' failed, due to generate a different schema, original schema: DFSchema { fields: [DFField { qualifier: Some(Bare { table: "lineitem" }), field: Field { name: "l_returnflag", data_type: Utf8, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} } }, DFField { qualifier: None, field: Field { name: "sum_disc_price", data_type: Decimal128(38, 2), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }], metadata: {} }, new schema: DFSchema { fields: [DFField { qualifier: Some(Bare { table: "lineitem" }), field: Field { name: "l_returnflag", data_type: Utf8, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} } }, DFField { qualifier: None, field: Field { name: "sum_disc_price", data_type: Decimal128(25, 2), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} } }], metadata: {} }. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
```

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #6819 #6790 .

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/tests/sqllogictests/test_files/pg_compat/pg_compat_null.slt [hunk 1]
```diff
@@ -70,7 +70,7 @@ WITH HEADER ROW
 LOCATION '../../testing/data/csv/aggregate_test_100.csv'
 
 
-statement ok
+statement error DataFusion error: Error during planning: Mismatch between schema and batches
```
Comment (jackwener) [inline datafusion/core/tests/sqllogictests/test_files/pg_compat/pg_compat_null.slt:73]: this error is due to `Mismatch between`

```
batches_schema
"Schema { fields: [Field { name: "c1", data_type: Utf8, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c2", data_type: Int8, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c3", data_type: Int16, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c4", data_type: Int16, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c5", data_type: Int32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c6", data_type: Int64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c7", data_type: Int16, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c8", data_type: Int32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c9", data_type: UInt64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c10", data_type: Utf8, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c11", data_type: Float32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c12", data_type: Float64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c13", data_type: Utf8, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "n5", data_type: Null, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "n9", data_type: UInt64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }"
str_batches_schema
"Schema { fields: [Field { name: "c1", data_type: Utf8, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c2", data_type: Int8, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c3", data_type: Int16, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c4", data_type: Int16, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c5", data_type: Int32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c6", data_type: Int64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c7", data_type: Int16, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c8", data_type: Int32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c9", data_type: UInt64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c10", data_type: Utf8, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c11", data_type: Float32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c12", data_type: Float64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "c13", data_type: Utf8, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "n5", data_type: Int32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "n9", data_type: UInt64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }"

after type_coercion, schema has changed.

```
name: "n5", data_type: Null, nullable: true
name: "n5", data_type: Int32,  nullable: true
```

**Comment by jackwener** [comment]
```
❯ SELECT "A", COUNT(DISTINCT("B"))
FROM "MyTable"
WHERE time >= now() - interval '1 month'

GROUP BY "A"



❯ SELECT "A", COUNT(DISTINCT("B"))
FROM "MyTable"
WHERE time >= now() - interval '1 month'
GROUP BY "A"
LIMIT 10;

+-----------+---------------------------+
| MyTable.A | COUNT(DISTINCT MyTable.B) |
+-----------+---------------------------+
+-----------+---------------------------+
0 rows in set. Query took 0.020 seconds.

```

**Review by alamb — COMMENTED** [review]
I think this needs some tests given that we didn't catch it initially, so we don't know if this code is covered or not.

I will contribute some shortly

**Comment by alamb** [comment]
BTW here are some tests (that I verified that this PR fixes the errors): https://github.com/apache/arrow-datafusion/pull/6836

**Review by alamb — APPROVED** [review]
Thanks @jackwener

**Comment by jackwener** [comment]
Thanks @alamb

**Pull request merged**